jormungandr/src/network/protocols/bootstrap.rs

```diff
@@ -0,0 +1,107 @@
+use super::super::{grpc, BlockConfig};
+use crate::blockcfg::Block;
+use crate::blockchain::protocols::{Blockchain, PreCheckedHeader, Ref};
+use crate::settings::start::network::Peer;
+use chain_core::property::HasHeader;
+use network_core::client::{block::BlockService, Client as _};
+use network_grpc::client::Connection;
+use slog::Logger;
+use std::fmt::Debug;
+use tokio::prelude::*;
+use tokio::runtime::current_thread;
+
+pub fn bootstrap_from_peer(peer: Peer, blockchain: Blockchain, tip: Ref, logger: &Logger) {
```

Comment:
so as you said in your comment in the PR, you are not sure about utilising a `Ref` here.

So indeed we will need the `Ref` so we know what is our current tip hash and we can actually query the peer. On another hand you also need to update the `Branch` you are building, continue to update it so when we are down the other task will know where to start from. Otherwise the `Tip` branch is still at your original `Tip` but not the updated (with this function) `Tip`.

So at the end this function needs to returns the last created `Ref` and then update the `Branch`.

so `bootstrap_from_stream` would return `Item = Ref` and then you would update the branch with this new `Ref`